engine/context/src/main/java/io/deephaven/engine/context/QueryCompiler.java

```diff
@@ -683,19 +683,51 @@ private void maybeCreateClass(String className, String code, String packageName,
         }
     }
 
+    private static volatile JavaCompiler JAVA_COMPILER = null;
+
+    private static JavaCompiler getJavaCompiler() {
+        JavaCompiler localCompiler;
+        if ((localCompiler = JAVA_COMPILER) == null) {
+            synchronized (QueryCompiler.class) {
+                if ((localCompiler = JAVA_COMPILER) == null) {
+                    localCompiler = JAVA_COMPILER = ToolProvider.getSystemJavaCompiler();
```

Comment:
One small improvement:

```java
JAVA_COMPILER = localCompiler = ...
```

does a volatile write, whereas the existing one does volatile write + read.